tiledb/api/src/range.rs

```diff
@@ -0,0 +1,841 @@
+                    }
+                    Ok(MultiValueRange::$V(value.1, value.2))
+                }
+            }
+        )+
+    }
+}
+
+multi_value_range_try_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+multi_value_range_try_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+multi_value_range_try_from!(Float32: f32, Float64: f64);
+
+#[macro_export]
+macro_rules! multi_value_range_go {
+    ($expr:expr, $DT:ident, $start:pat, $end:pat, $then:expr) => {
+        match $expr {
+            MultiValueRange::UInt8($start, $end) => {
+                type $DT = u8;
+                $then
+            }
+            MultiValueRange::UInt16($start, $end) => {
+                type $DT = u16;
+                $then
+            }
+            MultiValueRange::UInt32($start, $end) => {
+                type $DT = u32;
+                $then
+            }
+            MultiValueRange::UInt64($start, $end) => {
+                type $DT = u64;
+                $then
+            }
+            MultiValueRange::Int8($start, $end) => {
+                type $DT = i8;
+                $then
+            }
+            MultiValueRange::Int16($start, $end) => {
+                type $DT = i16;
+                $then
+            }
+            MultiValueRange::Int32($start, $end) => {
+                type $DT = i32;
+                $then
+            }
+            MultiValueRange::Int64($start, $end) => {
+                type $DT = i64;
+                $then
+            }
+            MultiValueRange::Float32($start, $end) => {
+                type $DT = f32;
+                $then
+            }
+            MultiValueRange::Float64($start, $end) => {
+                type $DT = f64;
+                $then
+            }
+        }
+    };
+}
+
+#[derive(Clone, Deserialize, Serialize, PartialEq)]
+pub enum VarValueRange {
+    UInt8(Box<[u8]>, Box<[u8]>),
+    UInt16(Box<[u16]>, Box<[u16]>),
+    UInt32(Box<[u32]>, Box<[u32]>),
+    UInt64(Box<[u64]>, Box<[u64]>),
+    Int8(Box<[i8]>, Box<[i8]>),
+    Int16(Box<[i16]>, Box<[i16]>),
+    Int32(Box<[i32]>, Box<[i32]>),
+    Int64(Box<[i64]>, Box<[i64]>),
+    Float32(Box<[f32]>, Box<[f32]>),
+    Float64(Box<[f64]>, Box<[f64]>),
+}
+
+impl VarValueRange {
+    pub fn check_datatype(&self, datatype: Datatype) -> TileDBResult<()> {
+        check_datatype!(self, datatype);
+        Ok(())
+    }
+}
+
+macro_rules! var_value_range_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl From<(Box<[$U]>, Box<[$U]>)> for VarValueRange {
+                fn from(value: (Box<[$U]>, Box<[$U]>)) -> VarValueRange {
+                    VarValueRange::$V(value.0, value.1)
+                }
+            }
+        )+
+    }
+}
+
+var_value_range_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+var_value_range_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+var_value_range_from!(Float32: f32, Float64: f64);
+
+impl From<(&str, &str)> for VarValueRange {
```

Comment:
FWIW it's not totally useless to have an additional `From<(String, String)>`, as that could avoid a copy by using `String::into_bytes`.

Obviously this is not a performance bottleneck but figured I'd mention.

Reply:
I'll add it. This is just an ergonomics API so that basically just us TileDB devs can write examples more easily. I didn't originally think someone would prefer to create a String instead of just writing `&["foo", "bar"]`, but looking again I could see it.

Reply:
[Add zero copy array VarValueRange helper](https://github.com/TileDB-Inc/tiledb-rs/pull/72/commits/854614b19fa7ead0a0892304c0ffd66ef976c2f4)